dsgrid/query/derived_dataset.py

```diff
@@ -0,0 +1,168 @@
+            table_records = get_unique_values(df, dim.model.dimension_query_name)
+        if is_valid:
+            dim_ref = project.config.get_dimension_reference(dim.model.dimension_id)
+            dimension_references.append(serialize_model(dim_ref))
+        else:
+            records = dim.get_records_dataframe()
+            records = records.filter(records.id.isin(table_records))
+            # TODO: AWS #186 - not an issue if registry is in a database instead of files
+            filename = dst_path / f"{dim.model.dimension_query_name}_records.csv"
+            # Use pandas because spark creates a directory.
+            records.toPandas().to_csv(filename, index=False)
+            # Use dictionaries to avoid validation.
+            new_dim = {
+                "type": dim.model.dimension_type.value,
+                "name": dim.model.name,
+                "display_name": dim.model.display_name,
+                "module": dim.model.module,
+                "class_name": dim.model.class_name,
+                "description": dim.model.description,
+                "filename": filename.name,
+            }
+            dimensions.append(new_dim)
+
+    config = {
+        "dataset_id": query.project.dataset.dataset_id,
+        "dataset_type": InputDatasetType.MODELED.value,
+        "data_source": "",
+        "data_schema_type": DataSchemaType.ONE_TABLE.value,
+        "data_schema": {
+            "load_data_column_dimension": metadata.pivoted.dimension_type.value,
+            "column_type": ColumnType.DIMENSION_QUERY_NAMES.value,
+        },
+        "dataset_version": "1.0.0",
+        "description": "",
+        "origin_creator": "",
+        "origin_organization": "",
+        "origin_date": "",
+        "origin_project": "",
+        "origin_version": "",
+        "source": "",
+        "data_classification": DataClassificationType.MODERATE.value,  # TODO
+        "use_project_geography_time_zone": True,
+        "dimensions": dimensions,
+        "dimension_references": dimension_references,
+    }
+    config_file = dst_path / DatasetRegistry.config_filename()
+    config_file.write_text(json5.dumps(config, indent=2))
+    logger.info(
+        "Created %s with default information. "
+        "Re-used %s project dimensions and generated %s new dimensions. "
+        "Examine %s, fill out the remaining fields, and review the generated dimension values "
+        "before submitting the dataset.",
+        dst_path,
+        len(dimension_references),
+        len(dimensions),
+        config_file,
+    )
+    return True
+
+
+def does_query_support_a_derived_dataset(query: ProjectQueryModel):
+    """Return True if a derived dataset can be created from a query.
+
+    Returns
+    -------
+    bool
+    """
+    is_valid = True
+    if query.result.supplemental_columns:
+        is_valid = False
+        logger.error("Cannot create a derived dataset from a query with supplemental_columns")
```

Comment:
I thought we can? And can register with the help of mappings?

Reply:
This is not the same thing as supplemental dimensions. Suppose you use county as the geography dimension but then request that state be added as a column. Now there are two geography dimensions, and so it is invalid.

Reply:
But this is not checking the number of cols for a given dim type, but just that supplemental_columns exist.

Reply:
That inherently means that multiple columns exist. It would be different if the user chose a supplemental dimension in the query.